mne/io/snirf/_snirf.py

```diff
@@ -531,49 +547,48 @@ def _get_lengthunit_scaling(length_unit):
         )
 
 
-def _extract_sampling_rate(dat):
+def _extract_sampling_rate(dat, user_sfreq):
     """Extract the sample rate from the time field."""
     # This is a workaround to provide support for Artinis data.
     # It allows for a 1% variation in the sampling times relative
     # to the average sampling rate of the file.
     MAXIMUM_ALLOWED_SAMPLING_JITTER_PERCENTAGE = 1.0
 
+    _validate_type(user_sfreq, ("numeric", None), "sfreq")
     time_data = np.array(dat.get("nirs/data1/time"))
-    sampling_rate = 0
-    if len(time_data) == 2:
-        # specified as onset, samplerate
-        sampling_rate = 1.0 / (time_data[1] - time_data[0])
+    time_unit = _get_metadata_str(dat, "TimeUnit")
+    time_unit_scaling = _get_timeunit_scaling(time_unit)  # always 1 (s) or 1000 (ms)
+    if len(time_data) == 2:  # special-cased in the snirf standard as (onset, period)
+        onset, period = time_data
+        file_sfreq = time_unit_scaling / period
     else:
-        # specified as time points
+        onset = time_data[0]
         periods = np.diff(time_data)
-        uniq_periods = np.unique(periods.round(decimals=4))
-        if uniq_periods.size == 1:
-            # Uniformly sampled data
-            sampling_rate = 1.0 / uniq_periods.item()
```

Comment:
this was also a bug. we shouldn't have used the rounded-to-4-decimals value. Now we use the mean of per-period sfreq estimates